libyul/optimiser/RedundantStoreEliminator.cpp

```diff
@@ -0,0 +1,452 @@
+#include <libyul/optimiser/KnowledgeBase.h>
+#include <libyul/ControlFlowSideEffectsCollector.h>
+#include <libyul/AST.h>
+
+#include <libsolutil/CommonData.h>
+
+#include <libevmasm/Instruction.h>
+#include <libevmasm/SemanticInformation.h>
+
+#include <range/v3/algorithm/all_of.hpp>
+
+using namespace std;
+using namespace solidity;
+using namespace solidity::yul;
+
+/// Variable names for special constants that can never appear in actual Yul code.
+static string const zero{"@ 0"};
+static string const one{"@ 1"};
+static string const thirtyTwo{"@ 32"};
+
+
+void RedundantStoreEliminator::run(OptimiserStepContext& _context, Block& _ast)
+{
+	map<YulString, SideEffects> functionSideEffects = SideEffectsPropagator::sideEffects(
+		_context.dialect,
+		CallGraphGenerator::callGraph(_ast)
+	);
+
+	SSAValueTracker ssaValues;
+	ssaValues(_ast);
+	map<YulString, AssignedValue> values;
+	for (auto const& [name, expression]: ssaValues.values())
+		values[name] = AssignedValue{expression, {}};
+	Expression const zeroLiteral{Literal{{}, LiteralKind::Number, YulString{"0"}, {}}};
+	Expression const oneLiteral{Literal{{}, LiteralKind::Number, YulString{"1"}, {}}};
+	Expression const thirtyTwoLiteral{Literal{{}, LiteralKind::Number, YulString{"32"}, {}}};
+	values[YulString{zero}] = AssignedValue{&zeroLiteral, {}};
+	values[YulString{one}] = AssignedValue{&oneLiteral, {}};
+	values[YulString{thirtyTwo}] = AssignedValue{&thirtyTwoLiteral, {}};
+
+	bool const ignoreMemory = MSizeFinder::containsMSize(_context.dialect, _ast);
+	RedundantStoreEliminator rse{
+		_context.dialect,
+		functionSideEffects,
+		ControlFlowSideEffectsCollector{_context.dialect, _ast}.functionSideEffectsNamed(),
+		values,
+		ignoreMemory
+	};
+	rse(_ast);
+	rse.changeUndecidedTo(State::Unused, Location::Memory);
+	rse.changeUndecidedTo(State::Used, Location::Storage);
+	rse.scheduleUnusedForDeletion();
+
+	StatementRemover remover(rse.m_pendingRemovals);
+	remover(_ast);
+}
+
+void RedundantStoreEliminator::operator()(FunctionCall const& _functionCall)
+{
+	RedundantStoreBase::operator()(_functionCall);
+
+	for (Operation const& op: operationsFromFunctionCall(_functionCall))
+		applyOperation(op);
+
+	ControlFlowSideEffects sideEffects;
+	if (auto builtin = m_dialect.builtin(_functionCall.functionName.name))
+		sideEffects = builtin->controlFlowSideEffects;
+	else
+		sideEffects = m_controlFlowSideEffects.at(_functionCall.functionName.name);
+
+	if (!sideEffects.canContinue)
+	{
+		changeUndecidedTo(State::Unused, Location::Memory);
+
+		changeUndecidedTo(
+			sideEffects.canTerminate ?
+			State::Used :
+			State::Unused,
+			Location::Storage
+		);
+	}
+}
+
+void RedundantStoreEliminator::operator()(FunctionDefinition const& _functionDefinition)
+{
+	ScopedSaveAndRestore storeOperations(m_storeOperations, {});
+	RedundantStoreBase::operator()(_functionDefinition);
+}
+
+
+void RedundantStoreEliminator::operator()(Leave const&)
+{
+	changeUndecidedTo(State::Used);
+}
```

Comment:
There should be a test case covering this case.

Reply:
Added it here: https://github.com/ethereum/solidity/pull/12311/files#diff-75043d51c2d8cfa2793d647366af90f8ce1b412f2470f2ebd80b0fdd432b7fdd